### PR TITLE
add e2e envs to devcontainer

### PR DESCRIPTION
**Pull request by ykadowak**
### Description:

This PR adds E2E environment to our VS Code devcontainer. With this devcontainer environment, we can just start k8s cluster with k3d, deploy vald, and run e2e without any additional operation.

Made the following changes:
- Changed devcontainer.json to build local Dockerfile because it's more flexible to customize dev environment for users and it does not take that much time to build anyway.
- Separate Dockerfile for devcontainer because it requires some hacky things only for VS Code.
- Enabled docker in docker feature of devcontainer to run E2E
- Removed devcontainer settngs and extensions from devcontainer.json because a VS Code user usually has their own custom settings for them so I think we don't need to force them to use specific settings.

### Versions:

- Go Version: 1.20.3
- Docker Version: 20.10.8
- Kubernetes Version: 1.22.0
- NGT Version: 2.0.11

### Checklist:

- [ ] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/main/CONTRIBUTING.md) document and completed [our CLA agreement](https://cla-assistant.io/vdaas/vald).
- [ ] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?

### CODE REVIEW

File: dockers/dev/README.md [hunk 1]
```diff
@@ -24,31 +24,18 @@ This image includes some libraries required for implementation and is based on `
...
 <!-- Vald Agent NGT requires more chapter Agent Standalone -->
 
-`vald-dev-container` is used for contributing to Vald's project.
-
-```bash
-docker pull vdaas/vald-dev-container
-docker run -name vald-dev -dit vdaas/vald-dev-container
-docker exec -it vald-dev /bin/bash
-```
+`vald-dev-container` is used for contributing to Vald's project. Please use this container with [our Dev Container settings](https://github.com/vdaas/vald/blob/main/.devcontainer/devcontainer.json).
```
Comment (vankichi) [inline dockers/dev/README.md:32]:
```suggestion
`vald-dev-container` is used for contributing to Vald's project.

Please use this container with [our Dev Container settings](https://github.com/vdaas/vald/blob/main/.devcontainer/devcontainer.json).
```

File: dockers/dev/Dockerfile [hunk 2]
```diff
@@ -64,12 +52,23 @@ COPY versions versions
...
     && make golangci-lint/install \
     && make reviewdog/install \
-    && make kubectl/install \
     && make protobuf/install
 
+# k3d
+SHELL ["/bin/bash", "-o", "pipefail", "-c"]
+RUN curl -sS https://raw.githubusercontent.com/k3d-io/k3d/main/install.sh | bash
+# kubectl
+RUN curl -LO "https://dl.k8s.io/release/$(curl -L -s https://dl.k8s.io/release/stable.txt)/bin/linux/amd64/kubectl"
```
Comment (kpango) [inline dockers/dev/Dockerfile:65]: I think it would be better to change Makefile.d/tools.mk#L74-L82

File: dockers/dev/README.md [hunk 3]
```diff
@@ -4,13 +4,13 @@
 
 `vald-dev-container` is designed for the development of Vald on Docker.
 
-This image includes some libraries required for implementation and is based on `ubuntu:devel` image.
+This image includes some libraries required to develop Vald and is based on `mcr.microsoft.com/vscode/devcontainers/go:1` image.
```
Comment (vankichi) [inline dockers/dev/README.md:7]: It would be better to show the specific URL of `mcr.microsoft.com/vscode/devcontainers/go:1`

**Review by vankichi — CHANGES_REQUESTED** [review]
`Dockerfile` and `devcontainer.json` seem to fine

File: dockers/dev/Dockerfile [hunk 4]
```diff
@@ -64,12 +52,23 @@ COPY versions versions
...
     && make valdcli/install \
     && make yq/install \
     && make golangci-lint/install \
     && make reviewdog/install \
-    && make kubectl/install \
     && make protobuf/install
 
+# k3d
+SHELL ["/bin/bash", "-o", "pipefail", "-c"]
+RUN curl -sS https://raw.githubusercontent.com/k3d-io/k3d/main/install.sh | bash
```
Comment (kpango) [inline dockers/dev/Dockerfile:63]: Also it would be better to change Makefile.d/k3d.mk#L27

**Review by kpango — CHANGES_REQUESTED** [review]

**Comment by kpango** [comment]
Others looks good to me.

File: Makefile.d/tools.mk [hunk 5]
```diff
@@ -77,7 +77,8 @@ $(BINDIR)/kubectl:
 	chmod a+x $(BINDIR)/kubectl
 else
 $(BINDIR)/kubectl:
-	curl -L "https://storage.googleapis.com/kubernetes-release/release/$(curl -s https://storage.googleapis.com/kubernetes-release/release/stable.txt)/bin/linux/amd64/kubectl" -o $(BINDIR)/kubectl
+	curl -LO "https://dl.k8s.io/release/$(shell curl -L -s https://dl.k8s.io/release/stable.txt)/bin/linux/amd64/kubectl"
+	install -o root -g root -m 0755 kubectl $(BINDIR)/kubectl
```
Comment (kpango) [inline Makefile.d/tools.mk:81]: I think user and permission is up to the user environment, how about using $USER for user and group definition
```suggestion
	install -o $USER -g $(id -g $USER) -m 0755 kubectl $(BINDIR)/kubectl
```

Comment (kpango) [reply Makefile.d/tools.mk:81]: And you can use it like
```
make USER=root kubectl/install
```

Comment (ykadowak) [reply Makefile.d/tools.mk]: Probably this is enough?

https://github.com/vdaas/vald/pull/2032/commits/1355abde2321898e64eeec07c6f402448d498c48

**Comment by kevindiu** [comment]
It works on my apple silicon mac including auto completion of cgo dependencies (e.g. internal/core/ngt tests) and e2e tests.
Will approve this PR after update branch

File: Makefile.d/tools.mk (hunk 5 above)
Comment (kpango) [reply Makefile.d/tools.mk]: Looks enough.

**Review by kpango — APPROVED** [review]
LGTM

**Comment by ykadowak** [comment]
@kevindiu 

> It works on my apple silicon mac including auto completion of cgo dependencies (e.g. internal/core/ngt tests) and e2e tests. Will approve this PR after update branch

By `apple silicon mac`, you mean opening VS Code devcontainer directly from mac os, not something like `lima`?  Which I didn't expect to work because of its archtecture.

**Comment by kevindiu** [comment]
@ykadowak yes, directly on macos with docker desktop.

**Review by kevindiu — APPROVED** [review]

**Review by vankichi — APPROVED** [review]
Thanks! LGTM :100:

**Pull request merged**